rfc/001-composable-pipelines/README.md

```diff
@@ -0,0 +1,353 @@
+
+```python
+import vast as v
+await v.from("s3://aws")
+       .read_json()
+       .group_by("src") # could also be merged with summarize()
+       .summarize(v.count("dst"))
+       .store_parquet("/path/to/directory")
+```
+
+### Synopsis
+
+The core of this proposal introduces pipeline operators as function with input
+and output types. The following overview introduces the pipeline building
+blocks used throughout this proposal.
+
+A pipeline operator has an input and output type, e.g., `from<In, Out>` takes as
+input `In` and produces instances of `Out`. We consider the following valid
+types:
+
+1. `Void`: no input or output, to represent a side effect
+2. `Bytes`: blocks of raw byte sequences
+3. `Arrow`: table slices (i.e., record batches)
+
+The majority of computation happens in (3), as this is the representation of
+structured data that the user primarily wants to interact with. The other types
+help in building a composable system to onboard and offboard data from the Arrow
+data plane.
+
+The central pipeline invariant is that composed operators must have a valid
+combination of types. For example, this is a valid pipeline:
+
+```
+from<Void, Bytes> |
+read<Bytes, Arrow> |
+where<Arrow, Arrow>
+```
+
+The dataflow edges (`|`) in the middle produce valid type pairs:
+
+- `<*, Void>`
+- `<Bytes, Bytes>`
+- `<Arrow, *>`
+
+The `*` character signals the open end of the pipeline. By collapsing the types
+in the middle, the type pipeline of the pipeline becomes `<Void, Arrow>`. Since
+`Void` does not require input to run, this pipeline could be "kicked off".
+However, there is consumer at the other end, as the pipeline ends in `Arrow`.
+Adding another operator would make it complete, e.g.:
+
+```
+from<Void, Bytes> |
+read<Bytes, Arrow> |
+where<Arrow, Arrow> |
+parquet<Arrow, Void>
+```
+
+This results in a `<Void, Void>` pipeline that does not "leak".
+
+With this typing concept in mind, we can now take a closer look at the various
+operators we propose.
+
+#### I/O
+
+Input:
+
+| Operator | Input Type | Output Type | Description
+| -------- | ---------- | ----------- | ------------------------------------------
+| `from`   | `Void`     | `Bytes`     | Loads bytes from a source as a side effect
+| `read`   | `Bytes`    | `Arrow`     | Parses a specific format into Arrow
+| `load`   | `Void`     | `Arrow`     | Performs `from` and `read` in one step
+
+Output:
+
+| Operator | Input Type | Output Type | Description
+| -------- | ---------- | ----------- | ------------------------------------------
+| `to`     | `Bytes`    | `Void`      | Writes bytes to a sink as a side effect
+| `write`  | `Arrow`    | `Bytes`     | Prints input in a specific format
+| `store`  | `Arrow`    | `Void`      | Performs `write` and `to` in one step
+
+The reason where `load`/`store` exist, in addition to `from`/`to` and
+`read`/`write`, is that it's not always possible to encapsulate data acquisition
```

Comment:
But then that poses the opposite question: Why do `from` and `to` exist?

If there are not always available for every plugin, the UX inconsistency may be more painful (I would need to check what each plugin I want to use supports). The only use case I could come up with to use `from` and `to` without `read` and `write` would be to copy raw full data sets from one place to the other, so maybe that could be done with a dedicated `copy<Void, Void>` operator?

Reply:
The raison d'etre for `from` and `to` is that it solves a composability. We noticed that there are two separate phases in data acquisition that users want to mix and match:

1. Loading data from some carrier. Examples:
  - Socket/file
  - Kafka
  - 0mq
  - RabbitMQ
  - Fluent
2. Parsing the (framed) data with a specific format
  - Zeek
  - JSON
  - CSV

If we had only `load`, we'd have to implement M * N combinations. With `from` and `read`, end up with M + N implementations.

In theory, `from` and `read` can be an implementation detail (of the physical representation), making only `load` the user-facing operator. Then you'd have to write `load - zeek` or `load kafka json`. What I like about this that the user doesn't have to burdened with thinking when to use `load` vs `from` + `read`.

On the UX side, this would mean that `load` takes one or two arguments, based on whether `from` and `read` are a single step. For example:
- `load broker`: get Zeek via Broker
- `load - zeek`: get Zeek via stdin

Would that be a better UX than spreading it over a total of 3 commands?

Reply:
I think so. By order of importance to me is:

1. Not having to learn too much about every plugin before starting to use it for a simple usage. I don't want to have to find what subset of `{from, read, to, write}` a given plugin supports. I don't mind if the implementation does match the UX, for example is from actually does from + read for Zeek's Broker plugin, as long as the output of the pipeline is the one intended.
2. Having fewer alternatives unless they have strong use-cases, because it's fewer things to learn and the tutorials will be easier to follow.

I think your proposal above takes care of both points. That said I can still be convinced otherwise, that wouldn't be  a blocker for me, but some more details to remember.

Reply:
I like this idea. The separation between carrier and format only is somewhat required to avoid the N x M combinations for the physical operators, but for the logical operators that is not actually a requirement. If we want to expose that to users as separate logical operators we can always do so in the future, but if we were to add them now we'd probably have a hard time removing them again if they turn out to be unnecessary.

Reply:
If implemented @netantho's suggestion in [8727b30](https://github.com/tenzir/vast/pull/2511/commits/8727b304f40d129b679aa263fd3b5e3d8b08f330).

Reply:
Closing out this thread, as it's taken care of. Feel free to re-open if there's open discussions left.